### PR TITLE
Bugfix: fix missing use of `oidc_token_file_path`, CLI auth with Cloud Shell

### PR DESCRIPTION
**Pull request by manicminer**
- **provider:** fix an issue with authentication using `oidc_token_file_path`
- **provider:** fix an issue with Azure CLI authentication when running in Azure Cloud Shell

Closes: https://github.com/hashicorp/terraform-provider-azurerm/issues/20674
Closes: https://github.com/hashicorp/terraform-provider-azurerm/issues/20794

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -14,7 +14,7 @@ require (
 	github.com/google/go-cmp v0.5.9
 	github.com/google/uuid v1.1.2
 	github.com/hashicorp/go-azure-helpers v0.55.0
-	github.com/hashicorp/go-azure-sdk v0.20230306.1165107
+	github.com/hashicorp/go-azure-sdk v0.20230307.1120859
```
Comment (tombuildsstuff) [inline go.mod:17]: there's some base layer changes included here, might be worth rebasing this once #20810 has been merged?

Comment (manicminer) [reply go.mod:17]: Gotcha, will do 👍

File: .github/workflows/provider-test.yml [hunk 2]
```diff
@@ -42,6 +42,9 @@ jobs:
         run: |
           echo "ARM_OIDC_TOKEN=$(curl -H "Accept: application/json; api-version=2.0" -H "Authorization: Bearer ${ACTIONS_ID_TOKEN_REQUEST_TOKEN}" -H "Content-Type: application/json" -G --data-urlencode "audience=api://AzureADTokenExchange" "${ACTIONS_ID_TOKEN_REQUEST_URL}" | jq -r '.value')"  >>${GITHUB_ENV}
 
+      - name: Set OIDC Token File Path
+        run: echo "${ARM_OIDC_TOKEN}" >"${TMPDIR}/oidc-token.jwt"
```
Comment (tombuildsstuff) [inline .github/workflows/provider-test.yml:46]: we'll want to move this down two steps, so this is only output if the other two are successful?

File: .github/workflows/provider-test.yml [hunk 3]
```diff
@@ -60,4 +63,7 @@ jobs:
           ARM_TENANT_ID: ${{ secrets.ARM_TENANT_ID }}
           ARM_SUBSCRIPTION_ID: ${{ secrets.ARM_SUBSCRIPTION_ID }}
 
+      - name: Clean Up OIDC Token File Path
```
Comment (tombuildsstuff) [inline .github/workflows/provider-test.yml:66]: at the moment this'll only run if it's successful, we can use an `always` for this to ensure it's always run: https://github.com/hashicorp/pandora/blob/6569798f7478abc0f7b23a120c13ac092e631fd3/.github/workflows/automation-regenerate-go-sdk.yaml#L57

**Review by tombuildsstuff — APPROVED** [review]
Left a couple of comments inline but 👍

**Pull request merged**